### DISCUSSION TITLE
Provide JLama categorization support

### DISCUSSION THREAD

**Discussion by lordofthejars** [quarkiverse/quarkus-langchain4j#1071] [Ideas]
If I am not wrong currently we are providing JLama support by using the LangChain4J integration through `ChatModel`. The problem is that JLama also supports classification feature, not just generating content. I have been using JLama in several examples, specially in GuardRails and the code most of the time looks like:

```java
 @ConfigProperty(name = "jailbreak.model", defaultValue = "lordofthejars/jailbreak-classifier")
    String jailbreakModel;
    static String workingDirectory = System.getProperty("user.home") + "/jlamamodels";

    File localJailbreakModelPath;

    @Startup
    public void init() throws IOException {
        this.localJailbreakModelPath = new Downloader(workingDirectory, jailbreakModel).huggingFaceModel();
    }

    @Produces
    @Named("jailbreak")
    AbstractModel createJailbreakModel() {
        return ModelSupport.loadClassifierModel(localJailbreakModelPath, DType.F32, DType.I8);
    }
````

And this is repeated for every classification model. So would it be possible to provide some integration where I set as quarkus property something like:

```properties
quarkus.langchain4j.jlama.jailbreak.model=lordofthejars/jailbreak-classifier
quarkus.langchain4j.jlama.download=true
quarkus.langchain4j.jlama.outputdir=....
```

Of course other properties could be download yes/no, download directory, .... optional with some defaults.

So then from my code the only thing I need to do is:

```java
@Named("jailbreak")
AbstractModel m;
```

So basically I help users to not have to repeat this code for every project and every model.

Maybe we could even provide some interface returning a Map and annotated with `@RegisterAiService`, or any other annotation but I think this would be more complicated.

**Comment by geoand** [comment]
This is interesting for sure.

Can you please link to a sample application that uses this idea?

**Reply by lordofthejars** [reply]
Oh yes: https://github.com/lordofthejars-ai/quarkus-langchain-examples/tree/main/prompt-injection-guard

I am changing this app as I am adding more guards, but the one it is released is this one for now.

**Reply by lordofthejars** [reply]
The producer: https://github.com/lordofthejars-ai/quarkus-langchain-examples/blob/main/prompt-injection-guard/src/main/java/org/acme/GuardModelsProducer.java

The injector: https://github.com/lordofthejars-ai/quarkus-langchain-examples/blob/main/prompt-injection-guard/src/main/java/org/acme/jailbreak/JailbreakModel.java

**Reply by geoand** [reply]
Thanks, I'll have a look soon

**Reply by lordofthejars** [reply]
Thanks mate, the idea is to be able to use more than one model in the same project, and we can use the `Named` annotation or any other annotation to inject the required model.

**Reply by geoand** [reply]
Right, I'll check if it makes sense to do something in the extension

**Reply by lordofthejars** [reply]
In fact and related to this PR, we could offer something like a `Classification` class where you can return the labels, or the best label according to probability or an ordered entries. 

`AbstractMap` returns just a Map so it is another thing I need to implement all the time, and we could provide the class. Apart from the class as utility, for example we could provide something like:


```java
@RegisterAi
@Named("jailbreak")
public interface Toxic {
  public Classification toxicity(String text);

}
```

So basically we are following the approach done in the other integrations but returning a Classification.

**Reply by geoand** [reply]
After looking at this I'm -1 for having the extension configure `AbstractModel`. If there are specific types that make sense to configure, we can consider it.

**Reply by lordofthejars** [reply]
ok, so you mean the classification stuff not implemented but the original question yes?

**Reply by geoand** [reply]
I am saying that if can come up with something that makes sense (the way `ChatLanguageModel` does), we can configure it. What we don't want to do is configure something totally generic.

**Reply by lordofthejars** [reply]
Ok, but then automatic creation of `AbstractModel` you are up to incorporate it?

**Reply by geoand** [reply]
No, that's what I am saying. There is going to be no automatic producing of `AbstractModel` - that's up to user code to do.

**Reply by lordofthejars** [reply]
Yes but then the code becomes really cumberstone as it is repeated everywhere. I think it is a good addition for improving dev experience on this, as we avoid user having to define all the Producer code every time with the configuration parameters and so on.

But it is ok, no worries. It was an idea after all

**Reply by geoand** [reply]
> Yes but then the code becomes really cumberstone as it is repeated everywhere.

That might be the case for these examples, but I see plenty of things that would have to be configurable to make this useful as something general, and this it doesn't seem to me at this point to be something that the extension should provide.
